### PR TITLE
refactor: make debug functionality better.

### PR DESCRIPTION
**Pull request by ayamir**
Dap for c/cpp/rust:
We use `lldb-vscode` to debug, but I found that it doesn't work on my arch wsl2 although I have set the `lldb-vscode`'s path to absolute path, ref: https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#ccrust-via-lldb-vscode
It needs more test on other platforms.
Dap for python:
It works now but needs user install `debugpy` manually.
ref: https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#Python

### CODE REVIEW

**Comment by ayamir** [comment]
I'm considering whether add an installation option for dap adapters like `cpptools` and `debugpy` in `install.sh` and `install.ps1`. Now I put the executable path in `~/.local/share/nvim/dap/`.

**Comment by CharlesChiuGit** [comment]
> Debug for c/cpp/rust can work when using `cpptools` instead.
> 
> I'm considering whether add an installation option for dap adapters like `cpptools` and `debugpy` in `install.sh` and `install.ps1`. Now I put the executable path in `~/.local/share/nvim/dap/`.
> 
> cc @Jint-lzxy @CharlesChiuGit 

I think that's a gd idea, to save some newcomers from confusing.

**Comment by Groveer** [comment]
> Dap for c/cpp/rust: 用于 c/cpp/rust 的 DAP： We use `lldb-vscode` to debug, but I found that it doesn't work on my arch wsl2 although I have set the `lldb-vscode`'s path to absolute path, ref: https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#ccrust-via-lldb-vscode 我们使用 `lldb-vscode` 进行调试，但我发现它在我的 arch wsl2 上不起作用，尽管我已将 `lldb-vscode` 的路径设置为绝对路径，参考：https://github.com/mfussenegger/nvim- dap/wiki/Debug-Adapter-installation#ccrust-via-lldb-vscode It needs more test on other platforms. 它需要在其他平台上进行更多测试。 Dap for python:  Dap 为 python： It works now but needs user install `debugpy` manually. 它现在可以工作，但需要用户手动安装 `debugpy` 。 ref: https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#Python


`lldb-vscode` works for me, and I do't need absolute path, I use env path.

图片
图片

File: lua/modules/configs/tool/dap/dap-lldb.lua [hunk 1]
```diff
@@ -1,25 +1,26 @@
 local dap = require("dap")
 
+-- https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#ccrust-via-lldb-vscode
+local lldb_abspath = vim.fn.exepath("lldb-vscode")
```
Comment (ClSlaid) [inline lua/modules/configs/tool/dap/dap-lldb.lua:4]: Can we also add codelldb (known as vscode-lldb, not lldb-vscode) for rust?
It has better pretty print support for rust.

Comment (ClSlaid) [reply lua/modules/configs/tool/dap/dap-lldb.lua:4]: related: https://github.com/ayamir/nvimdots/discussions/551

Comment (ayamir) [reply lua/modules/configs/tool/dap/dap-lldb.lua:4]: Currently working on my paper. Will impl it when I'm spare.

**Comment by ayamir** [comment]
screensht_23-06-12_01:18:07
screensht_23-06-12_01:28:23
screensht_23-06-12_01:28:54

**Review by aarnphm — COMMENTED** [review]
The DAP implementation LGTM. I used to use DAP a lot, not anymore. Still cc others to take a look

**Comment by Groveer** [comment]
LGTM. It works good for me.

**Review by CharlesChiuGit — APPROVED** [review]
LGTM, but I only tested on python

**Comment by Jint-lzxy** [comment]
Sorry for the delay. Will test on C++, Objective-C, and Swift.

**Review by Jint-lzxy — CHANGES_REQUESTED** [review]
LGTM, currently working on several cleanups like https://github.com/ayamir/nvimdots/pull/458#issuecomment-1411496895

**Comment by Jint-lzxy** [comment]
A small change in wording: Executables with LSP-like capabilities are called **"servers"**, while executables having DAP implementations are called **"clients"**.

**Review by Jint-lzxy — APPROVED** [review]
LGTM for `codelldb` and `lldb` 👍

**Comment by Jint-lzxy** [comment]
https://github.com/ayamir/nvimdots/pull/777/commits/1d9381cba5b3aaba9c66c8965eb16876801dd8e9 implemented a dirty workaround for `utils` (all parameters passed to the adapter should be functions or tables, and they shouldn't be called during setup):

```lua
return setmetatable({}, {
	__index = function(_, key)
		return function()
			return function()
				return M[key]()
			end
		end
	end,
})
```
Not sure if this is the best way to do so 😢

**Review by ClSlaid — APPROVED** [review]

**Comment by ayamir** [comment]
> all parameters passed to the adapter should be functions or tables, and they shouldn't be called during setup

You are right. Previous implementation has the bug about multi-times prompts asking for exe path.

**Comment by Jint-lzxy** [comment]
> We use `project.nvim` to change cwd automatically, so is it suitable to use `vim.fn.getcwd()` as default path completion? For example, I have a `Cpp-Practices` git repo to host my cpp code. Now I enter the directory of a leetcode problem and want to debug the executale. So what I needed is current directory path instead of the workspace path provided by `vim.fn.getcwd()`.

https://github.com/ayamir/nvimdots/pull/777/commits/fdb356c1514a4e8167d5ba5b3aaad737d654a537

**Pull request merged**